alf/environments/parallel_environment_test.py

```diff
@@ -100,6 +104,66 @@ def test_step(self):
                          time_step2.observation.shape)
         env.close()
 
+    def test_non_blocking_reset_with_spare_envs(self):
+        num_envs = 10
+        sleep_time = 1.
+        self._set_default_specs()
+
+        def slow_env_load(env_name, env_id=0):
+            return SlowStartingEnvironment(
+                self.observation_spec,
+                self.action_spec,
+                env_id=env_id,
+                max_duration=1,
+                time_sleep=sleep_time,
+                use_tensor_time_step=False)
+
+        start_time = time.time()
+        env = alf.environments.utils.create_environment(
+            env_name="IgnoredName",
+            env_load_fn=slow_env_load,
+            num_parallel_environments=num_envs,
+            start_serially=False,
+            num_spare_envs=num_envs)
+        init_time = time.time()
+        init_time_diff = init_time - start_time
+        self.assertLessEqual(
+            init_time_diff,
+            sleep_time * 2 - 0.5,
+            msg=('Expected all processes to start together, '
+                 'got {} wait time').format(init_time_diff))
+
+        time_step0 = env.reset()
+        action_spec = env.action_spec()
+        action = torch.stack([action_spec.sample() for _ in range(num_envs)])
+        time_step1 = env.step(action)
+        self.assertEqual(time_step0.observation.shape,
+                         time_step1.observation.shape)
+        step1_time = time.time()
+        time_step2 = env.step(action)
+        reset_time = time.time()
+        reset_time_diff = reset_time - step1_time
+        self.assertEqual(time_step1.observation.shape,
+                         time_step2.observation.shape)
+        self.assertLessEqual(
+            reset_time_diff,
+            sleep_time - 0.5,
+            msg=(f'Reset with spare envs took {reset_time_diff}, too long'))
+        time_step3 = env.step(action)
```

Comment:
Also need to test env_id is correct (i.e. from 0 to num_envs - 1)

Reply:
Done.